config/config_util.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"encoding/json"
+	"reflect"
+)
+
+// CloneConf ...
+func CloneConf(conf *Config) (*Config, error) {
+	content, err := json.Marshal(conf)
+	if err != nil {
+		return nil, err
+	}
+	var clonedConf Config
+	if err := json.Unmarshal(content, &clonedConf); err != nil {
+		return nil, err
+	}
+	return &clonedConf, nil
+}
+
+var (
+	// dynamicConfigItems contains all config items that can be changed during runtime.
+	dynamicConfigItems = []string{
+		"Performance.MaxProcs", "Performance.MaxMemory",
+		"Performance.CrossJoin", "Performance.FeedbackProbability",
+		"Performance.QueryFeedbackLimit", "Performance.PseudoEstimateRatio",
+		"OOMAction", "MemQuotaQuery", "TiKVClient.StoreLimit",
+	}
+)
+
+// MergeConfigItems is used to overwrite some config items that can't be changed during runtime.
+func MergeConfigItems(dstConf, newConf *Config) (acceptedItems, rejectedItems []string) {
+	return mergeConfigItemsWithAllowMap(dstConf, newConf, dynamicConfigItems)
+}
+
+func mergeConfigItemsWithAllowMap(dstConf, newConf *Config, dynamicConfigItems []string) (acceptedItems, rejectedItems []string) {
+	dynamicConfigItemsMap := make(map[string]struct{}, len(dynamicConfigItems))
+	for _, c := range dynamicConfigItems {
+		dynamicConfigItemsMap[c] = struct{}{}
+	}
+	return mergeConfigItems(reflect.ValueOf(dstConf), reflect.ValueOf(newConf), "", dynamicConfigItemsMap)
+}
+
+func mergeConfigItems(dstConf, newConf reflect.Value, fieldPath string, dynamicConfigItemsMap map[string]struct{}) (acceptedItems, rejectedItems []string) {
+	t := dstConf.Type()
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+		dstConf = dstConf.Elem()
+		newConf = newConf.Elem()
+	}
+	if t.Kind() != reflect.Struct {
+		if reflect.DeepEqual(dstConf.Interface(), newConf.Interface()) {
+			return
+		}
+		if _, ok := dynamicConfigItemsMap[fieldPath]; ok {
+			dstConf.Set(newConf)
+			return []string{fieldPath}, nil
+		}
+		return nil, []string{fieldPath}
+	}
+
+	for i := 0; i < t.NumField(); i++ {
+		p := t.Field(i).Name
+		if fieldPath != "" {
+			p = fieldPath + "." + p
+		}
+		as, rs := mergeConfigItems(dstConf.Field(i), newConf.Field(i), p, dynamicConfigItemsMap)
```

Comment:
I think that add a line `fieldPath = p` before line 80 maybe clearer.

Reply:
We can't do this since in the next iteration we need the original `fieldPath`.